server/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -234,19 +240,60 @@ private CompletionStage<Application> submitApplication(
     return applicationRepository
         .submitApplication(applicantId, programId, submitterEmail)
         .thenComposeAsync(
-            applicationMaybe -> {
+            (Optional<Application> applicationMaybe) -> {
               if (applicationMaybe.isEmpty()) {
                 return CompletableFuture.failedFuture(
                     new ApplicationSubmissionException(applicantId, programId));
               }
+
               Application application = applicationMaybe.get();
               String programName = application.getProgram().getProgramDefinition().adminName();
               notifyProgramAdmins(applicantId, programId, application.id, programName);
+
               if (submitterEmail.isPresent()) {
                 notifySubmitter(submitterEmail.get(), applicantId, application.id, programName);
               }
+
               maybeNotifyApplicant(applicantId, application.id, programName);
-              return CompletableFuture.completedFuture(application);
+
+              return updateStoredFileAclsForSubmit(applicantId, programId)
+                  .thenApplyAsync((ignoreVoid) -> application, httpExecutionContext.current());
+            },
+            httpExecutionContext.current());
+  }
+
+  /**
+   * When an application is submitted, we store the name of its program in the ACLs for each file in
+   * the application. This is so that when a program admin goes to view it we can authorize them,
+   * both for the original program the file was uploaded for as well as programs that the applicant
```

Comment:
I'd remove the original part and speak about how this works in general.  The "original" part is speaking towards legacy support that is otherwise not present or a concern in this class.

Reply:
> The "original" part is speaking towards legacy support that is otherwise not present or a concern in this class.

It's not speaking toward legacy support. "original" is talking about the first program the file was uploaded for. The entire paragraph his discussing new behavior.

Reply:
ah ok. Is that actually relevant for the new ACL model though? I don't think the order of programs the applicant applied too (and this is updated for) is relevant.

Reply:
I want to make sure it's clear why we have extra logic here for stored files vs all the other data that doesn't require this kind of ACLing behavior. Do you think the comment is adequate for that if I omit the second sentence?

Reply:
Yeah, I'd suggest just keeping the first sentence as it's directly pertinent to this code.  

When someone reads the code base, they don't know all the history, so they're going to see this as _the_ acl mechanism.  The legacy concerns are then documented and isolated to where they are relevant; and I believe can be removed in the near future via the TODOs you added.

Reply:
sounds good, done